src/unix/thread.c

```diff
@@ -2,6 +2,11 @@
 
 thread current;
 
+struct futex {
```

Comment:
This looks like a major improvement to the futex code. Would it be a pain to split it out into a futex.c? It's substantial enough and doesn't seem like something inseparably tied to the thread code.